accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+    duplicate keys in a map literal, according to behavior specified in the
+    runtime semantics.
+
+*   When evaluating a const collection literal, if there is an existing const
+    list, map, or set literal that contains the same series of values or entries
+    in the same order, then that object is used instead of producing a new one.
+    *In other words, constants are canonicalized.* Note that maps are only
+    canonicalized if they contain the same keys *in the same order*.
+
+### Constant equality safe
+
+A value is constant equality safe iff:
+
+*   It is an instance of a class that does not override `Object.operator==`, or
+
+*   it is an instance of `int` or `String`, or
+
+*   it is an instance of `Symbol` originally created by a symbol literal or a
+    constant invocation of the `Symbol` constructor, or
+
+*   it is an instance of `Type` originally created by evaluating a type literal
+    expression (an identifier or qualified identifier denoting a class, mixin or
+    type alias declaration, that is evaluated as an expression).
+
+All constant equality safe values are also valid constant values.
+
+## Dynamic Semantics
+
+Spread, `if`, and `for` behave similarly across all collection types. To
+simplify the spec, there is a single procedure used for all kinds of
+collections. This recursive procedure builds up either a `result` sequence of
+either values or key-value pair map entries. Then a final step handles those
+appropropriately for the given collection type.
+
+### To evaluate a collection `element`:
+
+1.  If `element` is an expression element:
+
+    1.  Evaluate the element's expression and append it to `result`.
+
+1.  Else, `element` is a `mapEntry` `keyExpression: valueExpression`:
+
+    1.  Evaluate `keyExpression` to a value `key`.
+
+    1.  Evaluate `valueExpression` to a value `value`.
+
+    1.  Append `key: value` to `result`.
+
+1.  Else, if `element` is a spread element:
+
+    1.  Evaluate the spread expression to a value `spread`.
+
+    1.  If `entry` is null-aware and `spread` is null, continue to the next
+        element in the literal.
+
+    1.  If the collection is a map, evaluate `spread.entries.iterator` to a
+        value `iterator`. Otherwise, evaluate `spread.iterator` to `iterator`.
+        *This will deliberately throw an exception if `spread` is `null` and
+        `element` is not null-aware.*
+
+    1.  Loop:
+
+        1.  If `iterator.moveNext()` returns `false`, exit the loop.
+
+        1.  Evaluate `iterator.current` and append it to `result`. *This will be
+            a MapEntry in a map literal, or any object for a list or set
+            literal.*
+
+1.  Else, if `element` is an `if` element:
+
+    1.  Evaluate the condition expression to a value `condition`.
+
+    1.  If the boolean conversion of `condition` is `true`:
+
+        1.  Evaluate the "then" element using this procedure.
+
+    1.  Else, if there is an "else" element of the `if`:
```

Comment:
'Else, if `condition` is `true`..'.

And then 'A dynamic error occurs if `condition` is the null object' in a separate case.